### ISSUE TITLE
MongoError: insertDocument :: caused by :: 11000 E11000 duplicate key error index: test.attachments.locks.$files_id_1  dup key: { : null }

### ISSUE THREAD

**Issue by sorcererlc** [vsivsi/gridfs-locking-stream#4]
Hi,

This is more of a call for help than an issue, as i've been seeing this error for the past 5 days and my headache level reached critical today :)

I don't know if this is an issue with the module or with my code, but i've checked everything and Google doesn't seem to be of any help at all in this case.

This used to work before updating to v1.x. No changes were made to my code.

From what i see in the module, files_id should never be null if i specify _id in the options object, which i do, but it crashes when creating a new LockCollection, at [index.js:53](https://github.com/vsivsi/gridfs-locking-stream/blob/master/index.js#L53).

My code is as follows:

```javascript
var gfs = new Grid(mongoose.connection.db, mongoose.mongo, collection),
		options = {
			_id: new mongoose.Types.ObjectId(),
			mode: 'w',
			filename: attachmentDetails.filename,
			content_type: attachmentDetails.mimeType,
			root: 'attachments',
			metadata: {}
		},
		result = {
			sequence_id: options._id,
			file_name: options.filename,
			mime_type: options.content_type,
			size: buffer.length,
			metadata: {}
		};

// Meta processing stuffs

gfs.createWriteStream(options, function (err, writeStream) {
		if (err) {
			logHelper.error('db', new Error(err), options);

			return callback(err);
		}

		if (writeStream) {
			var dataStream = streamifier.createReadStream(buffer);

			dataStream.pipe(writeStream);

			return callback(null, result);
		} else {
			return callback({
				code: 500,
				message: 'A write lock could not be obtained'
			})
		}
	});
```

Do you see anything wrong with it?

Thanks in advance for your help and thanks for a great module.

**Comment by vsivsi** [comment]
Hi, thanks for your question. At first glance, I suspect this issue has to do with the fact that version 1.x of this package now requires the MongoDB connection to bottom out in the new version 2.0.x version of the native node.js mongodb driver.  This is a requirement that bubbled up from the gridfs-stream package, and the change was necessary because it was the only way to support new-style node.js 0.10+ streams. (The old mongodb 1.4.x native node.js driver only supports old style node 0.8.x streams).  Anyway, these changes are why the version of this package and the gridfs-stream package each bumped to version 1.x, because this transition is not entirely backwards compatible (e.g. gridfs file append is no longer supported...)

So getting back to your sample code, if the mongoose connection you are passing is using the mongodb 1.4.x driver internally, then that is almost certainly the cause of your issue.

If this is the problem, then this package (and gridfs-stream) should do a better job of detecting this case and throwing an early error on instantiation when an unsupported mongodb driver is in use.

You should be able to see which version of the mongodb driver is in use in your app by by looking at the package file for the version of mongoose you have installed. I'm not a big mongoose user, but I assume it specifies which mongodb driver it uses there.

Please let me know what you find, and I hope this helps.

**Comment by sorcererlc** [comment]
Hi,

After trying to revert to v0.2.6 of gridfs-locking-stream, installing the native MongoDB driver v2.0 and creating a separate connection for file storage i kept getting the same error.

Then I tried backtracking through the module's code and see exactly where the error occurs. This led me to [index.js:24](https://github.com/vsivsi/gridfs-locks/blob/master/index.js#L24) in the gridfs-locks module.

The problem seems to be with isMongoDriver20() when multiple queries are queued (my app makes a lot of parallel queries to the database). The test document seems to be inserted but not removed before another gfs.createWriteStream() call is made.

**Comment by vsivsi** [comment]
Hi, thanks for investigating this further.  You are correct, I've been able to replicate this and there is a race condition in that function when multiple clients create lock collections backed by the same database at the same time.  I'm working on a fix now.

**Comment by vsivsi** [comment]
Hi, I just published v1.3.2 of gridfs-locks and v1.0.2 of this package to fix this issue.  

I'm a bit troubled that there seems to be no standard way to distinguish connection objects from the mongodb 1.x and 2.x node.js drivers.  I have a question out to the mongodb package maintainer to see if there's a more reliable way to tell the difference.  If I get a good response, I'll will update these packages again to reflect that method.

**Comment by sorcererlc** [comment]
Hi, glad I could help. This seems to solve my issue.

Thanks, again, for a great module.
